keps/sig-apps/4650-stateful-set-update-claim-template/README.md

```diff
@@ -0,0 +1,1087 @@
+     Only update the PVC when the Pod is deleted.
+   
+4. When patching the PVC, if we also re-create the Pod,
+   update the PVC after old Pod deleted, together with creating new pod.
+   Otherwise, if pod is not changed, update the PVC only.
+
+Failure cases: don't left too many PVCs being updated in-place. We expect to update the PVCs in order.
+
+- If the PVC update fails, we should block the update process.
+  If the Pod is also deleted (by controller or manually), don't block the creation of new Pod.
+  We should retry and report events for this.
+  The events and status should look like those when the Pod creation fails.
+
+- While waiting for the PVC to reach the compatible state,
+  We should update status, just like what we do when waiting for Pod to be ready.
+  We should block the update process if the PVC is never compatible.
+
+- If the `volumeClaimTemplates` is updated again when the previous rollout is blocked,
+  similar to [Pods](https://kubernetes.io/docs/concepts/workloads/controllers/statefulset/#forced-rollback),
+  user may need to manually deal with the blocking PVCs (update or delete them).
+
+
+### What PVC is compatible
+
+A PVC is compatible with the template if:
+- All the immutable fields match exactly; and
+- `metadata.labels` and `metadata.annotations` of PVC is a superset of the template; and
+- `status.capacity.storage` of PVC is greater than or equal to
+  the `spec.resources.requests.storage` of the template; and
+- `status.currentVolumeAttributesClassName` of PVC is equal to
+  the `spec.volumeAttributesClassName` of the template.
+
+### User Stories (Optional)
+
+<!--
+Detail the things that people will be able to do if this KEP is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system. The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+#### Story 1: Batch Expand Volumes
+
+We're running a CI/CD system and the end-to-end automation is desired.
+To expand the volumes managed by a StatefulSet,
+we can just use the same pipeline that we are already using to update the Pod.
+All the test, review, approval, and rollback process can be reused.
+
+#### Story 2: Shinking the PV by Re-creating PVC
+
+After running our app for a while, we optimize the data layout and reduce the required storage size.
+Now we want to shrink the PVs to save cost.
+We can not afford any downtime, so we don't want to delete and recreate the StatefulSet.
+We also don't have the infrastructure to migrate between two StatefulSets.
+Our app can automatically rebuild the data in the new storage from other replicas.
+So we update the `volumeClaimTemplates` of the StatefulSet,
+delete the PVC and Pod of one replica, let the controller re-create them,
+then monitor the rebuild process.
+Once the rebuild completes successfully, we proceed to the next replica.
+
+#### Story 3: Asymmetric Replicas
+
+The storage requirement of different replicas are not identical,
```

Comment:
I've wondered about the `volumeClaimUpdatePolicy` usefulness, but this particular usecase nicely describes when using `OnDelete` value for that field will be useful. So you can ignore my previous question about it.